### PR TITLE
Compute MergedSelections as a memoized tree

### PR DESCRIPTION
**Pull request by AnthonyMDev**
Memoizes computation of merged selections using `MergedSelectionTree`. This requires keeping track of type scopes for all parent types in the path to the selection set.

This PR enables merged selection computation across nested child selection sets across an entire operation.

### CODE REVIEW

File: Sources/ApolloCodegenLib/Frontend/GraphQLType.swift [hunk 1]
```diff
@@ -2,24 +2,37 @@ import JavaScriptCore
...
   case nonNull(GraphQLType)
   case list(GraphQLType)
 
   public var typeReference: String {
     switch self {
-    case let .named(type):
+    case let .entity(type as GraphQLNamedType),
+         let .scalar(type as GraphQLNamedType),
+         let .enum(type as GraphQLNamedType),
+         let .inputObject(type as GraphQLNamedType):
```
Comment (designatednerd) [inline Sources/ApolloCodegenLib/Frontend/GraphQLType.swift:17]: Ooh fancy, I didn't know you could do this if all the associated objects are the same type

File: Sources/ApolloCodegenLib/IR/ScopedSelectionSetHashable.swift [hunk 2]
```diff
@@ -13,38 +13,21 @@ protocol ScopedSelectionSetHashable {
...
-
-extension CompilationResult.Field: ScopedSelectionSetHashable {
-  var hashForSelectionSetScope: String {
-    return responseKey
-  }
-}
-
-extension ASTField: ScopedSelectionSetHashable {
+extension IR.SelectionSet: ScopedSelectionSetHashable {
+#warning("What if there is a field with the same name as a type? Do we get a conflict? Write a test for this.")
```
Comment (designatednerd) [inline Sources/ApolloCodegenLib/IR/ScopedSelectionSetHashable.swift:17]: Adding a comment as a flag to check that we're testing this later (and for @AnthonyMDev to put a link to the commit with that test 😇)

File: Sources/ApolloCodegenTestSupport/MockIRModels.swift [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+//      name,
+//      alias: alias,
+//      arguments: arguments,
+//      type: .scalar(type),
+//      selectionSet: selectionSet)
+//
+//    return IR.Field(mockField, enclosingEntityMergedSelectionBuilder: nil)
+//  }
+//}
+//
```
Comment (designatednerd) [inline Sources/ApolloCodegenTestSupport/MockIRModels.swift:38]: do you still need all this commented out stuff?

File: Sources/ApolloUtils/LinkedList.swift [hunk 4]
```diff
@@ -0,0 +1,185 @@
+/// A doubly linked list implementation.
```
Comment (designatednerd) [inline Sources/ApolloUtils/LinkedList.swift:1]: l33t h4x0rz!!1! 

I'm kind of surprised there isn't anything in the stdlib or any of the helper libs around it - might be worth contributing?

File: Sources/ApolloUtils/LinkedList.swift [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  }
+
+  public func makeIterator() -> Iterator {
+    Iterator(self)
+  }
+}
+
+extension LinkedList: CustomDebugStringConvertible where T: CustomDebugStringConvertible {
+  public var debugDescription: String {
+    "[head] -> \(headNode.debugDescription)"
```
Comment (designatednerd) [inline Sources/ApolloUtils/LinkedList.swift:177]: do you only want to log out the head here or the head with all the nodes?

File: Sources/ApolloUtils/ResponsePath.swift [hunk 6]
```diff
@@ -0,0 +1,76 @@
+/// Represents a list of string components joined into a path using a reverse linked list.
+///
+/// A response path is stored as a linked list because using an array turned out to be
+/// a performance bottleneck during decoding/execution.
```
Comment (designatednerd) [inline Sources/ApolloUtils/ResponsePath.swift:4]: out of curiosity, how bad was the bottleneck?

File: Tests/ApolloCodegenTests/AnimalKingdomAPI/AnimalKingdomASTCreationTests.swift [hunk 7]
```diff
@@ -12,61 +12,71 @@ final class AnimalKingdomASTCreationTests: XCTestCase {
...
+              type: .nonNull(.entity(GraphQLObjectType.mock("Height")))),
         .mock("species",
-              type: .nonNull(.named(GraphQLScalarType.string()))),
+              type: .nonNull(.scalar(GraphQLScalarType.string()))),
         .mock("skinCovering",
-              type: .named(GraphQLEnumType.skinCovering())),
+              type: .enum(GraphQLEnumType.skinCovering())),
         .mock("predators",
-              type: .nonNull(.list(.nonNull(.named(Interface_Animal))))),
+              type: .nonNull(.list(.nonNull(.entity(Interface_Animal))))),
```
Comment (designatednerd) [inline Tests/ApolloCodegenTests/AnimalKingdomAPI/AnimalKingdomASTCreationTests.swift:79]: The ability to tell what these are in terms of entity vs scalar vs enum is aaawesome

**Review by designatednerd — COMMENTED** [review]
looking good! few notes and questions.

File: Sources/ApolloUtils/LinkedList.swift (hunk 4 above)
Comment (designatednerd) [reply Sources/ApolloUtils/LinkedList.swift:1]: (Thinking specifically of [`swift-collections`](https://github.com/apple/swift-collections))

Comment (calvincestari) [reply Sources/ApolloUtils/LinkedList.swift:1]: IIRC there were optimizations in this implementation that made it specific to our performance needs.

File: Tests/ApolloCodegenTests/AnimalKingdomAPI/ExpectedGeneratedOutput/Queries/AllAnimalsQuery.swift [hunk 8]
```diff
@@ -276,7 +297,7 @@ public struct AllAnimalsQuery: GraphQLQuery {
 
         public var isJellicle: Bool { data["isJellicle"] }
         public var bodyTemperature: Int { data["bodyTemperature"] }
-        public var height: Height { data["height"] }
+        public var height: AllAnimal.AsPet.AsWarmBlooded.Height { data["height"] }
```
Comment (calvincestari) [inline Tests/ApolloCodegenTests/AnimalKingdomAPI/ExpectedGeneratedOutput/Queries/AllAnimalsQuery.swift:300]: Is this the intended change? I thought after we talked yesterday we decided not to go this route?

File: Tests/ApolloCodegenTests/CodeGenAST/IROperationBuilderTests.swift [hunk 9]
```diff
@@ -0,0 +1,3031 @@
...
+    // when
+    buildSubjectOperation()
+
+    let allAnimals = self.subject[field: "query"]?[field: "allAnimals"]?.selectionSet
+
+    // then
+    expect(allAnimals?.selections.typeCases).to(beEmpty())
+  }
+
+  /// Example:
```
Comment (calvincestari) [inline Tests/ApolloCodegenTests/CodeGenAST/IROperationBuilderTests.swift:75]: These comments-as-examples are super helpful for test readability. I think we can improve them slightly though:
* include relevant SDL for the operation
* enclose in "```" top and bottom so they're more readable in Xcode

This one would end up looking like..
```
  /// ```
  /// SDL:
  /// type Query {
  ///   allAnimals: [Animal!]!
  /// }
  /// interface Animal {
  ///   species: String!
  /// }
  /// type Bird implements Animal {
  ///   species: String!
  /// }
  ///
  /// OPERATION:
  /// query {
  ///  allAnimals {
  ///    ...BirdDetails
  /// }
  ///
  /// fragment BirdDetails on Bird {
  ///   species
  /// }
  /// Expected:
  /// AllAnimals.children: [AsBird]
  /// ```
```

File: Tests/ApolloCodegenTests/CodeGenAST/IROperationBuilderTests.swift [hunk 10]
```diff
@@ -0,0 +1,3031 @@
...
+  /// Children should not include a type case for asAnimal
+  func test__children__isObjectType_initWithNamedFragmentOnLessSpecificMatchingType_hasNoChildTypeCase() {
+    // given
+    let Interface_Animal = GraphQLInterfaceType.mock("Animal")
+    let Object_Bird = GraphQLObjectType.mock("Bird", interfaces: [Interface_Animal])
+    let animalDetails = CompilationResult.FragmentDefinition.mock("AnimalDetails", type: Interface_Animal)
+
+    operation = .mock(selections: [
+      .field(.mock(
+        "allAnimals",
```
Comment (calvincestari) [inline Tests/ApolloCodegenTests/CodeGenAST/IROperationBuilderTests.swift:135]: Should this be `"birds"`? `[field: "allAnimals"]` would need to change down below too.

**Review by calvincestari — APPROVED** [review]
Approving this while still getting through the test cases. The example comments need revision to ensure it matches with what the tests are doing.

**Comment by calvincestari** [comment]
@AnthonyMDev this is great work - thank you! I've resolved a couple of the comments and left others that we should sync up on when you're back. I'm going to merge this now and we can revisit changes in future PRs, if needed. This crucially fixes test builds and my next pieces of work need to layer on top of this.

**Pull request merged**